Support for a remote signing via a TCP proxy

dancvrcek opened the pull request:
Hi

can you please consider this PR introducing a support for remote signing via a TCP proxy - the motivation is to support high-security signing devices like smartcards, or HSMs.

The changes introduce a new keyStoreType - "REMOTE". A specification of the keyStore is done with the standard command line options:
-  keyStoreFile option, which expects a string with a hostname and a port, e.g., "localhost:4444",
- keyStorePassword - provides a PIN/password to unlock the remote device (optional); and
- keyStoreAlias - identification of the key - e.g., a name in a certificate (similar to my-windows keystore)

The code changes of the pull request consist of two main parts:
1. changes to KeyStoreUtils.java - add an "abstract" keystore type "REMOTE" and pulling aliases for the REMOTE keystore from the remote TCP proxy
2. the actual signing in SignerLogic.java 

The signing changes required 2 blocs of a new code: 1. to obtain certificates, and 2. to get the actual signature.

The protocol to the proxy is fairly simple, and it's described in more detailed at https://gitlab.com/cloudfoxy/foxy_tcpproxy

This pull requests uses 3 commands:
### ALIASES

An initial and optional command, which returns a list of all keyStoreAliases = names from certificates.
```
>all readers|
>1111:ALIASES|
```

which returns a list of all names in one line, encoded with base64, and separated with "|".

### CHAIN

This a request for all a certificate chain for a given keyStoreAlias.
```
>my keystoreAlias|
>1112:CHAIN|
```

which returns a list of certificates - base64 encoded and separated with "|".

### SIGN

The actual signing request.

```
>my keystoreAlias|
>1113:SIGN:<message hash>|
```

The response to the is an RSA signature, which can be added to the PdfPKCS7 as an external digest.

kwart commented on src/net/sf/jsignpdf/SignerLogic.java:
```diff
@@ -132,9 +143,67 @@ public boolean signFile() {
+			final PrivateKey key;
+			final Certificate[] chain;
+			// we first load certificates
+			if (this.options.getKsType().equals("remote")){
```
The keystore name has to be more descriptive. E.g. FoxyProxy or sth like it.

kwart commented on src/net/sf/jsignpdf/SignerLogic.java:
```diff
@@ -132,9 +143,67 @@ public boolean signFile() {
+						chain[i - 1] = certFactory.generateCertificate(in);
+					}
+				} catch (UnknownHostException ex) {
+					LOGGER.error("Can't connect to signing server" + hostname +": " + ex.getMessage());
```
The messages should be handled through resources, so the translations are possible.

kwart commented on src/net/sf/jsignpdf/SignerLogic.java:
```diff
@@ -361,7 +430,75 @@ public boolean signFile() {
-			sgn.update(sh, 0, sh.length);
+
+			// THIS IS THE SIGNING, we need to have a new branch for external signers
+            if (this.options.getKsType().equals("remote")){
```
Wrong indentation type.

dancvrcek commented:
Hi @kwart 

thanks so much for the first positive response and feedback, it's much appreciated! 

We were looking into different options how to integrate with Java crypto API, but it resulted in a fairly complicated integration. That's why we came up with the idea of a TCP proxy, which is flexible enough to connect to external crypto providers.

In terms of the integration into the signer logic - what would be an acceptable approach? Would a class with the new KS with a couple of methods, let's say `getChain(alias)` and `getSignature(alias, hash)` ... something like that work? ... a third method could be `getKeyAliases()` for keyStoreUtils.java .

not sure if that's what you had in mind.

kwart commented:
You understood the separation well. I imagine a separate class or few classes in a separate package.

dancvrcek commented:
I have created an interface and a class in a new package extcsp and reverted all (at least I tried) redundant changes to SignerLogic and KeyStoreUtils.

The CSP is also renamed to a more specific "CloudFoxy". 

Please, have a look if it makes sense.

kwart commented on src/net/sf/jsignpdf/SignerLogic.java:
```diff
@@ -132,9 +133,24 @@ public boolean signFile() {
+			// the 'cloudfoxy' crypto provider computes signatures externally and there are no
+			// certificates or keys available via Java CSPs -> they have to be pulled from an
+			// external source in 2 steps: 1. certificate chain, 2. signature itself
+			if (this.options.getKsType().toLowerCase().equals("cloudfoxy")){
```
Could you extract the `"cloudfoxy"` to a constant?
Also the `getKsType()` may return `null`, so this would throw an NPE.

kwart commented on src/net/sf/jsignpdf/SignerLogic.java:
```diff
@@ -361,7 +377,19 @@ public boolean signFile() {
-			sgn.update(sh, 0, sh.length);
+
+			// THIS IS THE SIGNING, we need to have a new branch for external signers
+			if (this.options.getKsType().equals("cloudfoxy")) {
```
dtto: Constant+null-safety, sth like
```
CLOUDFOXY.equals(options.getKsType())
```

dancvrcek commented:
I used a StringUtils method - would it do?

kwart commented:
I think we are ready to merge. Could you add a new line describing your changes to the `doc/ChangeLog.txt` and squash the commits to a single one?
